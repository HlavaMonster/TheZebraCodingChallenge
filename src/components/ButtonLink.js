import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { SVGPATHS } from './Constants';

class ButtonLink extends Component {
  constructor() {
    super();
    this.handleButtonClick = this.handleButtonClick.bind(this);
  }

  handleButtonClick(link) {
    window.open(link, '_blank');
  }

  render() {
    var buttonClass = classNames('btn', {
      'btn-primary': this.props.isPrimary,
    });

    return (
      <a href={this.props.buttonAction} className={buttonClass} target={this.props.newTab && '_blank'}>
        <span>
          {this.props.buttonIcon &&
            this.props.buttonIcon !== '' && (
              <span className="btn-icon" dangerouslySetInnerHTML={{ __html: SVGPATHS[this.props.buttonIcon] }} />
            )}
          <span>{this.props.buttonText}</span>
        </span>
      </a>
    );
  }
}

ButtonLink.propTypes = {
  buttonAction: PropTypes.string,
  isPrimary: PropTypes.bool,
  buttonIcon: PropTypes.string,
  buttonText: PropTypes.string,
};

export default ButtonLink;
