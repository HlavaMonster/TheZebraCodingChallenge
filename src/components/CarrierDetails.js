import React from 'react';

const FeaturesList = props => {
  return <ul>{props.features.map((feature, index) => <li key={index}>{feature.name}</li>)}</ul>;
};

const CarrierDetails = props => {
  return (
    <div className="carrier-details">
      {(props.featuresHtml || (props.features && props.features.length > 0)) && (
        <div className="carrier-details-section carrier-features">
          <div className="detail-header">Features</div>
          {props.featuresHtml ? (
            <div className="detail-body" dangerouslySetInnerHTML={{ __html: props.featuresHtml }} />
          ) : (
            <div className="detail-body">
              <FeaturesList features={props.features} />
            </div>
          )}
        </div>
      )}
      {props.detailBody &&
        props.detailBody.length && (
          <div className="carrier-details-section carrier-description">
            <div className="detail-header">Why {props.name}?</div>
            <div className="detail-body">{props.detailBody}</div>
          </div>
        )}
    </div>
  );
};

export default CarrierDetails;
