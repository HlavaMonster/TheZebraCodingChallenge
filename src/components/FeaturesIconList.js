import React from 'react';
import { SVGPATHS } from './Constants';

const getSVGforIconClass = iconClass => {
  return SVGPATHS[iconClass];
};

const FeaturesIconList = props => {
  return (
    <ul className="features-icon-list">
      {props.features.map((feature, i) => (
        <li key={i} dangerouslySetInnerHTML={{ __html: getSVGforIconClass(feature.icon) }} />
      ))}
    </ul>
  );
};

export default FeaturesIconList;
