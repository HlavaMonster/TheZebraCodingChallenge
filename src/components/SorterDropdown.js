import React, { Component } from 'react';
import PropTypes from 'prop-types';

class SorterDropdown extends Component {
  render() {
    return (
      <div className="sorter">
        <span>Sort by:</span>
        <select value={this.props.value} onChange={this.props.changeHandler}>
          <option value="best">Best matches</option>
          <option value="rating">Rating (stars)</option>
          <option value="alphabetical">Alphabetical</option>
          <option value="price">Price</option>
        </select>
      </div>
    );
  }
}

SorterDropdown.propTypes = {
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  changeHandler: PropTypes.func.isRequired,
};

export default SorterDropdown;
