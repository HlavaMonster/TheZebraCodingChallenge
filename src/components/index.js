export { default as CarrierCard } from './CarrierCard';
export { default as CarrierCardList } from './CarrierCardList';
export { default as SorterDropdown } from './SorterDropdown';
export { default as SVGPATHS } from './Constants';
export { default as ButtonLink } from './ButtonLink';
