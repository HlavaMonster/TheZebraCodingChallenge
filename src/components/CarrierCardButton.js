import React from 'react';
import PropTypes from 'prop-types';
import ButtonLink from './ButtonLink';

const CarrierCardButton = props => {
  if (props.buttonType === 2) {
    return null;
  }

  let isPrimary = props.buttonType === 3;

  return (
    <ButtonLink
      buttonAction={props.buttonAction}
      buttonIcon={isPrimary ? 'IconBoltWhite' : 'IconArrowDown'}
      buttonText={props.buttonText || 'Buy Now'}
      isPrimary={isPrimary}
      newTab={true}
    />
  );
};

CarrierCardButton.propTypes = {
  buttonAction: PropTypes.string,
  buttonText: PropTypes.string,
  buttonType: PropTypes.number,
};

export default CarrierCardButton;
