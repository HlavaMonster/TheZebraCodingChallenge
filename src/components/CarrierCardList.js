import React, { Component } from 'react';
import sortBy from 'lodash.sortby';
import { CarrierCardContainer } from 'containers';
import { SorterDropdown } from 'components';

class CarrierCardList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      sorter: 'best',
    };
    this.handleSortingChange = this.handleSortingChange.bind(this);
  }

  handleSortingChange(event) {
    this.setState({
      sorter: event.target.value,
    });
  }

  getSorterProperty() {
    let sorter = '';
    switch (this.state.sorter) {
      case 'alphabetical':
        sorter = 'card_order_a_to_z';
        break;
      case 'rating':
        sorter = 'stars';
        break;
      case 'price':
        sorter = 'card_order_price';
        break;
      case 'best':
      default:
        sorter = 'card_order_best_match';
        break;
    }
    return sorter;
  }

  render() {
    return (
      <div className="carrier-card-list">
        <div className="search-summary-section">
          <div className="search-summary">
            You just compared <span>{this.props.carriersSearched} rates</span> in{' '}
            {Math.ceil(this.props.searchTime / 1000 / 60)} minutes!
          </div>
          <div className="sorting">
            <SorterDropdown val={this.state.sortOrder} changeHandler={this.handleSortingChange} />
          </div>
        </div>
        {sortBy(this.props.carrierCards, card => {
          return card[this.getSorterProperty()];
        }).map((card, index) => <CarrierCardContainer key={index} carrierCard={card} />)}
      </div>
    );
  }
}

export default CarrierCardList;
