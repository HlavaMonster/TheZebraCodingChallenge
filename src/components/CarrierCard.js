import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import CarrierCardButton from './CarrierCardButton';
import CarrierDetails from './CarrierDetails';
import CornerTag from './CornerTag';
import StarRating from './StarRating';
import FadeAndSlide from './TransitonFadeSlide';
import CarrierLogo from '../assets/logo-carrier-sm-default.png';
import FeaturesIconList from './FeaturesIconList';

const CarrierRate = props => {
  return (
    <div className="carrier-rate">
      <div className="rate-tag">{props.ratetag}</div>
      <div className="rate">
        <span className="currency">$</span>
        <span className="value">{props.rate && props.rate.toFixed(0)}</span>
        <span className="term">/ mo</span>
      </div>
    </div>
  );
};

class CarrierCard extends Component {
  render() {
    var carrierCardClass = classNames('carrier-card', {
      'carrier-card-expanded': this.props.isExpanded,
    });
    return (
      <div className={carrierCardClass} onClick={this.props.onClick}>
        {this.props.cornerTag && <CornerTag />}
        <div className="carrier-summary">
          <div className="carrier-info">
            <div className="carrier-logo">
              <img src={CarrierLogo} alt="carrier logo" height="60" />
            </div>
            <div>
              <div className="carrier-name">{this.props.name}</div>
              <div className="carrier-tagline">{this.props.tagline}</div>
              <div className="carrier-features-rating">
                <StarRating stars={this.props.stars} />
                {this.props.features && <FeaturesIconList features={this.props.features} />}
              </div>
            </div>
          </div>
          <div className="carrier-rate-info">
            {this.props.rate && <CarrierRate rate={this.props.rate} ratetag={this.props.ratetag} />}
            <CarrierCardButton
              buttonText={this.props.actionLinkText}
              buttonType={this.props.type}
              buttonAction={'http://www.' + this.props.actionLink}
            />
          </div>
        </div>
        <FadeAndSlide in={this.props.isExpanded} timeout={500} mountOnEnter={true} unmountOnExit={true}>
          <div>
            <CarrierDetails
              name={this.props.name}
              featuresHtml={this.props.featuresHtml}
              features={this.props.features}
              detailBody={this.props.detailBody}
              isExpanded={this.props.isExpanded}
            />
          </div>
        </FadeAndSlide>
        <div className="toggle-icon">
          <svg width="11.67" height="6.04" viewBox="0 0 11.67 6.04">
            <polygon points="5.89 6.04 0 0.74 0.67 0 5.88 4.69 11 0 11.67 0.74 5.89 6.04" fill="#959595" />
          </svg>
        </div>
      </div>
    );
  }
}

CarrierCard.propTypes = {
  name: PropTypes.string,
  tagline: PropTypes.string,
  ratetag: PropTypes.string,
  rate: PropTypes.number,
  isExpanded: PropTypes.bool,
  actionLink: PropTypes.string,
  actionLinkText: PropTypes.string,
  type: PropTypes.number.isRequired,
  stars: PropTypes.number,
  detailBody: PropTypes.any,
  features: PropTypes.array,
  featuresHtml: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
  onClick: PropTypes.func,
  cornerTag: PropTypes.bool,
};

export default CarrierCard;
