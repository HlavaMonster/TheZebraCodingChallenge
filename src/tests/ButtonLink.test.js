import React from 'react';
import ReactDOM from 'react-dom';
import { ButtonLink } from 'components';
import renderer from 'react-test-renderer';

let standardButton = <ButtonLink buttonAction="" buttonText="Test Button" buttonIcon="" />;
let buttonWithIcon = <ButtonLink buttonAction="" buttonText="Test Button" buttonIcon="IconTestIcon" />;
let primaryButton = <ButtonLink buttonAction="" buttonText="Test Button" buttonIcon="" isPrimary={true} />;

it('renders standard button without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(standardButton, div);
});

it('renders primary button without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(primaryButton, div);
});

it('renders correct standard button type', () => {
  const standardBtn = renderer.create(standardButton);
  let button = standardBtn.toJSON();
  expect(button).toMatchSnapshot();
});

it('renders correct primary button type', () => {
  const primaryBtn = renderer.create(primaryButton);
  let button = primaryBtn.toJSON();
  expect(button).toMatchSnapshot();
});

it('renders button icon if supplied', () => {
  const iconBtn = renderer.create(buttonWithIcon);
  let button = iconBtn.toJSON();
  expect(button).toMatchSnapshot();
});
