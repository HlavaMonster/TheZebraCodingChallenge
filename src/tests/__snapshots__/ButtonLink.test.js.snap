// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders button icon if supplied 1`] = `
<a
  className="btn"
  href=""
  target={undefined}
>
  <span>
    <span
      className="btn-icon"
      dangerouslySetInnerHTML={
        Object {
          "__html": undefined,
        }
      }
    />
    <span>
      Test Button
    </span>
  </span>
</a>
`;

exports[`renders correct primary button type 1`] = `
<a
  className="btn btn-primary"
  href=""
  target={undefined}
>
  <span>
    
    <span>
      Test Button
    </span>
  </span>
</a>
`;

exports[`renders correct standard button type 1`] = `
<a
  className="btn"
  href=""
  target={undefined}
>
  <span>
    
    <span>
      Test Button
    </span>
  </span>
</a>
`;
